### ISSUE TITLE
Test 640 is flaky

### ISSUE THREAD

**Issue by bboreham** [weaveworks/weave#2340]
E.g.

https://circleci.com/gh/weaveworks/weave/6122
https://circleci.com/gh/weaveworks/weave/6130
https://circleci.com/gh/weaveworks/weave/6147

````
>>> Test ./640_proxy_restart_reattaches_test.sh finished after 85.8 secs with error: exit status 1
Proxy restart reattaches networking to containers
bfb351603213d95645ee06b906040d1489c5bda95eca6a0d1aa998856927e665
43034cda5c47137cd129faeac6681a094f54a6cd64f20f1ea948ea5abf9db103
c2
Waiting for c2 on host2-6147-0.us-central1-a.positive-cocoa-90213 to be attached
weave
Waiting for c2 on host2-6147-0.us-central1-a.positive-cocoa-90213 to be attached
Waiting for c2 on host2-6147-0.us-central1-a.positive-cocoa-90213 to be attached
Waiting for c2 on host2-6147-0.us-central1-a.positive-cocoa-90213 to be attached
host2-6147-0.us-central1-a.positive-cocoa-90213> docker stop/waiting
host2-6147-0.us-central1-a.positive-cocoa-90213> docker start/running, process 12718
Waiting for proxy to start
Waiting for proxy to start
Waiting for c2 on host2-6147-0.us-central1-a.positive-cocoa-90213 to be attached
Waiting for c2 on host2-6147-0.us-central1-a.positive-cocoa-90213 to be attached
Waiting for c2 on host2-6147-0.us-central1-a.positive-cocoa-90213 to be attached
c71185f4afb5df4f49158ae5dfa48616b519eb37facebf46d118a064675f2e38
Waiting for c3 on host2-6147-0.us-central1-a.positive-cocoa-90213 to be attached
c3
unable to open process 21665 namespace: no such file or directory
````

**Comment by rade** [comment]
We are seeing this quite a lot.

https://circleci.com/gh/weaveworks/weave/6367

is another recent example.

**Comment by brb** [comment]
Most likely https://github.com/weaveworks/weave/blob/c720ebec02308a2e89eee359087b4e03f9e2da85/test/640_proxy_restart_reattaches_test.sh#L54 fails.

One hypothesis is the following race: after starting router we call `populate_router` which gets container addresses via `containerAddrs` of `weaveutil`. In the latter we do 1) check if a container is alive 2) get a container netdev (this can fail with "unable to open process X namespace: no such file or directory"). So, if a container dies after 1) but before 2) we get the error you see in the log. Why it can happen? Before re-lauching router we do `run_many` which starts many short-lived containers and _maybe_ they are responsible for triggering the race.

**Comment by awh** [comment]
Fixed by 71e43c2023228b6eacccc9955af662874c8c61f0
